### ISSUE TITLE
How to work with database connections across processes?

### ISSUE THREAD

**Issue by demitri** [ray-project/ray#13938]
I'm opening a new issue, but this is the same question as #10778.

I am new to Ray, so I may have some of the concepts incorrect.  I want to use Ray to process a large number of entries from an external database, e.g. PostgreSQL. Database connections (e.g. via SQLAlchemy) are not typically serializable - they are scoped to the process they are created in and not applicable outside of that process. A db session or SQLAlchemy object will break when serialized and created in another process.

The typical way to get around this problem is to pass the object's (integer) ID to the multiprocessing code and then on the receiving end, query the database with a process-local session to get the object. The issue to avoid is creating and destroying `n` sessions to manage `n` objects. Ideally, one session would be created per process and then reused for every task processed by that session.

Is it possible to do this in Ray? Initialize and reuse an object to be scoped locally to a single process?

**Comment by demitri** [comment]
Bump.

**Comment by fwong03** [comment]
I'm also new to Ray and found this issue while trying to do the same thing (one connection per process). Since this issue was opened a while ago @demitri I'm guessing you probably found a solution by now. But just in case this is helpful for anyone else who was in my position and googling around, I ended up getting these helpful two suggestions from Eric Liang in the Ray slack channel and I've copy and pasted his suggestions below:

1. Using Datasets:
> This is pretty similar to the GPU-state-setup problem for batch inference. Datasets has a pattern that supports this well

2. Creating an actor:
> Otherwise, you could use actors and do the connection in the actor constructor.

I ended up doing the second suggestion, creating an actor class with a method that's the task I want to run. Then I created an actor pool (https://docs.ray.io/en/master/actors.html#actor-pool) so there would be multiple actors available to work on the tasks in parallel. It seems to work, and hopefully this is actually doing what I intended!

**Comment by andreapiso** [comment]
Would it be possible to have a code snippet for this? Seems like distributing the data coming from a dB across multiple Ray nodes would be a really common requirement

**Comment by fwong03** [comment]
@andreapiso I reread the opening comment and I misunderstood the request, what I had been looking to do was run a bunch of tasks in parallel while limiting the number of database connections created, i.e., to reuse database connections when I could. 

Just in case this is what you are also trying to do (and not share database connections between processes as the original reuqest was), a pattern that I've found super helpful is to create a pool of actors and call `map_unordered` (or `map` if you care about order) like this:

```
import ray
from ray.util import ActorPool


@ray.remote
class ActorThatQueries:
    def __init__(self):
        # configure db connection stuff here

    def query_db(self, val):
        # code that queries db 
        return res

actors = [ActorThatQueries.remote() for _ in range(5)]
pool = ActorPool(actors)

# Get 100 things done but it's limited to max of 5 running at a time, so you limit the number of db connections you're using
results = list(pool.map_unordered(fn=lambda a, v: a.query_db.remote(v), values=list(range(100))))
```

**Comment by wjrforcyber** [comment]
> I'm opening a new issue, but this is the same question as #10778.
> 
> I am new to Ray, so I may have some of the concepts incorrect. I want to use Ray to process a large number of entries from an external database, e.g. PostgreSQL. Database connections (e.g. via SQLAlchemy) are not typically serializable - they are scoped to the process they are created in and not applicable outside of that process. A db session or SQLAlchemy object will break when serialized and created in another process.
> 
> The typical way to get around this problem is to pass the object's (integer) ID to the multiprocessing code and then on the receiving end, query the database with a process-local session to get the object. The issue to avoid is creating and destroying `n` sessions to manage `n` objects. Ideally, one session would be created per process and then reused for every task processed by that session.
> 
> Is it possible to do this in Ray? Initialize and reuse an object to be scoped locally to a single process?

So any chance show us how did you solve it ? Thanks.

**Comment by rdabane** [comment]
@demitri 
How did you solve this issue?
I'm also interesting in this.

**Comment by demitri** [comment]
@rdabane I never really adopted Ray so unfortunately I can't help!

**Comment by BitPhinix** [comment]
There is also always the option to spawn a pgbouncer / similar and let that handle the pooling for you. Often times that is a lot simpler
